### PR TITLE
Add documentation, doc-tests and examples to the secret-memory crate.

### PR DESCRIPTION
**Pull request by DavidNiehues**

### CODE REVIEW

**Comment by DavidNiehues** [comment]
> @DavidNiehues Can you add some more unit tests to drive up the coverage?

Sure! I took a close look at the codecov report. I see potential for actual higher coverage in `public.rs` and `secret.rs`. According to codecov, there still some uncovered functions in `malloc.rs`, `memfdsec.rs` and `alloc/mod.rs` that actually have doc-tests right above them that cover them but aren't detected as such by codecov.

**Comment by DavidNiehues** [comment]
@koraa I now added more unit tests for `secret.rs` and `public.rs`.

**Comment by DavidNiehues** [comment]
@koraa @aparcar I made the same style change around doc-tests with the `?`-operator and rebased onto main here as well.

File: secret-memory/src/secret.rs [hunk 1]
```diff
@@ -47,37 +48,47 @@ where
...
         if let Some(inner) = &mut self.0 {
-            let inner: &mut SecretBox<T> = inner; // type annotation
+            let inner: &mut SecretBox<T> = inner; // Type annotation.
             inner.zeroize()
         }
     }
 }
 
 impl<T: Zeroize + ?Sized> Drop for ZeroizingSecretBox<T> {
+    /// Relaases the memory of this [ZeroizingSecretBox]. In contrast to usual implementations
```
Comment (aparcar) [inline secret-memory/src/secret.rs:89]: Typo

File: secret-memory/src/secret.rs [hunk 2]
```diff
@@ -145,22 +159,50 @@ impl SecretMemoryPool {
...
+///
+/// // We have to define the security policy before using Secrets.
+/// secret_policy_use_only_malloc_secrets();
+/// let mut my_secret: Secret<32> = Secret::zero();
+/// // Fill with some random data that I can use a cryptographic key later on.
+/// my_secret.randomize();
+/// // In case I accidentally print my secret in a debug, it's still not leaked:
+/// assert_eq!(format!("{:?}", my_secret), "<SECRET>");
+/// // The data will be zeroized either way in the end, but for good measure, let's zeroize anymay
+/// my_secret.zeroize();
```
Comment (aparcar) [inline secret-memory/src/secret.rs:188]: If it's randomized anyway, should the documentation motivate users to actively do it? This seems to add extra randomization steps where there are not needed. Don't we trust our automatic zeroing step?

File: secret-memory/src/secret.rs [hunk 3]
```diff
@@ -200,16 +244,25 @@ impl<const N: usize> Randomize for Secret<N> {
...
     fn zeroize(&mut self) {
         if let Some(inner) = &mut self.storage {
             inner.zeroize()
         }
     }
 }
 
+// Our implementation of Drop gives back the allocated secret memory to the secret memory pool.
 impl<const N: usize> Drop for Secret<N> {
+    /// Relaases the memory of this [Secret]. In contrast to usual implementations to [Drop] we
```
Comment (aparcar) [inline secret-memory/src/secret.rs:260]: Typo

File: secret-memory/src/secret.rs [hunk 4]
```diff
@@ -200,16 +244,25 @@ impl<const N: usize> Randomize for Secret<N> {
     }
 }
 
+// Indicate that a Secret is always zeroized when it is dropped.
```
Comment (aparcar) [inline secret-memory/src/secret.rs:247]: Missing brackets around Secret?

**Comment by aparcar** [comment]
Thanks again for your work, I think this is mostly ready except some minor things!

**Review by aparcar — CHANGES_REQUESTED** [review]
Please fix the remaining typos and I'm happy to merge this.

**Review by aparcar — APPROVED** [review]

**Pull request merged**